### ISSUE TITLE
redundant closure lint fails for FnMut where FnOnce is expected

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust-clippy#3898]
This compiles ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7e067bffb234f1b39857d6dd986e1313)):

```rust
pub fn bar(mut x: Box<dyn FnMut()>) {
    foo(|| x())
}

pub fn foo<T: FnOnce()>(_: T) {}
```

but clippy suggests:

```
warning: redundant closure found
 --> src/lib.rs:2:9
  |
2 |     foo(|| x())
  |         ^^^^^^ help: remove closure as shown: `x`
  |
  = note: #[warn(clippy::redundant_closure)] on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```

which fails to compile with ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bd910962f25b592909b97a16c5198686)):

```
error[E0277]: expected a `std::ops::FnOnce<()>` closure, found `std::boxed::Box<dyn std::ops::FnMut()>`
 --> src/lib.rs:2:5
  |
2 |     foo(x)
  |     ^^^ expected an `FnOnce<()>` closure, found `std::boxed::Box<dyn std::ops::FnMut()>`
  |
  = help: the trait `std::ops::FnOnce<()>` is not implemented for `std::boxed::Box<dyn std::ops::FnMut()>`
  = note: wrap the `std::boxed::Box<dyn std::ops::FnMut()>` in a closure with no arguments: `|| { /* code */ }
note: required by `foo`
 --> src/lib.rs:5:1
  |
5 | pub fn foo<T: FnOnce()>(_: T) {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

**Comment by mati865** [comment]
cc https://github.com/rust-lang/rust-clippy/issues/1439

**Comment by g-bartoszek** [comment]
I'll try to fix it.
